rclpy/rclpy/__init__.py

```diff
@@ -129,6 +129,23 @@ def shutdown(*, context: Context = None, uninstall_handlers: Optional[bool] = No
         uninstall_signal_handlers()
 
 
+def try_shutdown(*, context: Context = None) -> None:
+    """
+    Try shutting down a previously initialized context.
+
+    This will also shutdown the global executor if the context was successfuly shutdown.
+
+    :param context: The context to invalidate. If ``None``, then the default context is used
+        (see :func:`.get_default_context`).
+    """
+    global __executor
+    _try_shutdown(context=context)
+    if not ok():
+        if __executor is not None:
+            __executor.shutdown()
+            __executor = None
```

Comment:
I wish the default executor registered an `on_shutdown()` callback with the default Context and handled itself that way. Would you be willing to try that? If not, I'd approve this change as it does seem to make the two shutdown functions more alike.

Reply:
That seems to be working correctly, see https://github.com/ros2/rclpy/pull/868/commits/cb23bd7827a6e69c7348021d53dc1877bd9fc705